dcos_launch/gce.py

```diff
@@ -0,0 +1,94 @@
+import json
+import logging
+
+from dcos_launch import util
+from dcos_test_utils import gce
+
+log = logging.getLogger(__name__)
+
+TEMPLATE = """
+- type: compute.v1.instance
+  name: {name}
+  properties:
+    zone: {zone}
+    machineType: zones/{zone}/machineTypes/{machineType}
+    disks:
+    - deviceName: boot
+      type: PERSISTENT
+      boot: true
+      autoDelete: true
+      initializeParams:
+        sourceImage: projects/coreos-cloud/global/images/family/{sourceImage}
+    networkInterfaces:
+    - network: global/networks/default
+      # Access Config required to give the instance a public IP address
+      accessConfigs:
+      - name: External NAT
+        type: ONE_TO_ONE_NAT"""
+
+
+class BareClusterLauncher(util.AbstractLauncher):
+    # Launches a homogeneous cluster of plain GMIs intended for onprem DC/OS
+    def __init__(self, config):
+        credentials_path = util.set_from_env('GOOGLE_APPLICATION_CREDENTIALS')
+        node_count = 1 + config['num_masters'] + config['num_public_agents'] + config['num_private_agents']
+        with open(credentials_path, 'r') as creds:
+            self.gce_wrapper = gce.GceWrapper(json.load(creds))
+
+        all_template = 'resources:'
+        for i in range(node_count):
+            all_template += TEMPLATE.format(name='vm' + str(i), sourceImage=config['sourceImage'],
+                                            machineType=config['machineType'], zone=config['zone'],
+                                            project_id=self.gce_wrapper.project_id)
+        config['template_body'] = all_template
+
+        self.config = config
+        self._deployment = gce.Deployment(self.gce_wrapper, config['deployment_name'], config['zone'])
+        self.insert_request_errors = None
+
+    def get_deployment(self):
+        if self._deployment is None:
+            raise Exception('Deployment creation request failed with: ' + str(self.insert_request_errors))
+        if self._deployment.errors is not None:
```

Comment:
From what I can tell, `self._deployment.errors` is only set by Deployment.get_info which is only called by Deployment.wait. This makes it very confusing because now this will only do anything if wait is called. So, when you go to do delete, you don't have any guarantee that this deployment exists.

In general, keeping hard errors in state like this will make more issues, rather try to raise these errors immediately